python/lsst/ts/standardscripts/base_close_loop.py

```diff
@@ -0,0 +1,499 @@
+            filter=self.filter,
+        )
+
+        # Set visit id
+        visit_id = int(image[0])
+
+        # Run WEP
+        self.mtcs.mtaos.cmd_runWEP.set_start(visitId=visit_id)
+
+    async def compute_ofc_offsets(self) -> None:
+        """Compute offsets using ts_ofc."""
+
+        # Create the config to run OFC
+        config = {
+            "rotation_angle": self.rotation_angle,
+            "filter_name": self.filter,
+            "comp_dof_idx": {
+                "m2HexPos": self.used_dofs[:5],
+                "camHexPos": self.used_dofs[5:10],
+                "M1M3Bend": self.used_dofs[10:30],
+                "M2Bend": self.used_dofs[30:],
+            },
+        }
+        config_yaml = yaml.dump(config, default_flow_style=False)
+
+        # Run OFC
+        await self.mtcs.mtaos.cmd_runOFC.set_start(config=config_yaml)
+
+        # Return offsets
+        return await self.mtcs.mtaos.evt_degreeOfFreedom.next(
+            flush=False, timeout=STD_TIMEOUT
+        )
+
+    async def arun(self, checkpoint: bool = False) -> None:
+        """Perform wavefront error measurements and DOF adjustments until the
+        thresholds are reached.
+
+        Parameters
+        ----------
+        checkpoint : `bool`, optional
+            Should issue checkpoints
+
+        Raises
+        ------
+        RuntimeError:
+            If coordinates are malformed.
+        """
+
+        for i in range(self.max_iter):
+            self.log.debug(f"Closed Loop iteration {i + 1} starting...")
+
+            if checkpoint:
+                await self.checkpoint(
+                    f"[{i + 1}/{self.max_iter}]: Closed Loop loop starting..."
+                )
+
+                await self.checkpoint(f"[{i + 1}/{self.max_iter}]: Taking image...")
+
+            # Run the operational mode handler function.
+            await self.operation_model_handlers[self.mode]()
+
+            # Save the wavefront error
+            wavefront_error = await self.mtcs.mtaos.evt_wavefrontError.next(
+                flush=False, timeout=STD_TIMEOUT
+            )
+
+            self.log.info(
+                f"Wavefront error zernike coefficients: {wavefront_error} in um."
+            )
+
+            # Compute ts_ofc offsets
+            dof_offset = await self.compute_ofc_offsets()
+
+            # Check if corrections have converged. If they have, then we stop.
+            if all(abs(dof_offset) < self.threshold):
+                self.log.info(f"OFC offsets are inside tolerance ({self.threshold}). ")
+                if checkpoint:
+                    await self.checkpoint(
+                        f"[{i + 1}/{self.max_iter}]: Closed Loop converged."
+                    )
+
+                self.log.info("Closed Loop completed successfully!\n")
+                return
+
+            # If we haven't converged and apply_corrections is true,
+            # then we apply the corrections
+            elif self.apply_corrections:
```

Comment:
I think you always want to apply the correction, even if you already converged. This way we don't trow away any iteration. I'd move this check to be before the convergence criteria check and make them independent `if` statements, e.g.;

```
if self.apply_corrections:
    ...

if all(abs(dof_offset) < self.threshold):
    ...
```

you don't need the `else` step.

Also, if we are not applying the correction, do we want to continue taking data? we might want to break or return in that case, e.g.;

```
if self.apply_corrections:
    ...
else:
    return
```

Reply:
I think we do want to keep taking data if we are not applying the correction. The reason being that in that way we can see how the closed loop corrections evolve without applying them. We might be repositioning the telescope or checking the LUT while doing this

Reply:
Ok, that's a good point. In this case, please, add a note in the `max_iter` description explaining that, if `apply_corrections` is `False` it will take this many pairs of images.